### DISCUSSION TITLE
Adding a component to newly spawned entities

### DISCUSSION THREAD

**Discussion by Sycrosity** [bevyengine/bevy#8827] [Q&A] [answered]
I've been having some trouble adding a component whenever a new entity is spawned, so that I can remove all entities with this component when switching to the MainMenu state - I first searched around for a bevy `Event` or similar that would allow inbuilt functionality for this, but came up short.
I attempted to fix this with two systems: one which adds the component (`Loaded`) to all entities `Without<Loaded>`, and another which `OnExit` of the state would remove all components `With<Loaded>`:

```rust
fn add_loaded_component(
    mut commands: Commands,
    query: Query<Entity, (Without<Loaded>, Without<Persist>)>,
) {
    for entity in &query {
        commands.entity(entity).insert(Loaded);
    }
}

pub fn teardown<T: Component>(
    to_despawn: Query<Entity, (With<T>, Without<Persist>)>,
    mut commands: Commands,
) {
    for entity in &to_despawn {
        commands.entity(entity).despawn_recursive();
    }
}
```
(Persist is so that e.g. the camera wouldn't be removed)

However, upon running a plugin using these:
```rust
impl Plugin for GalaxyWorldPlugin {
    fn build(&self, app: &mut App) {
        app.add_systems((
            ///other stuff...///
            add_loaded_component.in_set(OnUpdate(EngineState::InGame)),
            teardown::<Loaded>.in_schedule(OnExit(EngineState::InGame)),
        ));
    }
}
```

I kept recieving the error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NoEntities("bevy_ecs::query::state::QueryState<&bevy_egui::EguiContext, bevy_ecs::query::filter::With<bevy_window::window::PrimaryWindow>>")', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy-inspector-egui-0.18.3/src/quick.rs:80:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'Compute Task Pool (0)' panicked at 'A system has panicked so the executor cannot continue.: RecvError', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_ecs-0.10.1/src/schedule/executor/multi_threaded.rs:194:60
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_tasks-0.10.1/src/task_pool.rs:376:49
thread 'Compute Task Pool (1)' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.10.1/src/pipelined_rendering.rs:111:77
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_tasks-0.10.1/src/task_pool.rs:376:49
```

and cannot work out what went wrong or how I should have approached this instead. Is there a more bevy-like way of approaching this, and if not what went wrong here?

**Comment by wilk10** [comment, marked as answer]
What happened is that you added the `Loaded` component to internal Bevy entities as well, specifically the entity with the `PrimaryWindow` component in this case, and then the `teardown` system has despawned them, but those are needed and should be managed by Bevy. 

You'll need a solution that marks what should be despawned more granularly. You could add the `Loaded` component at the time of spawning your own entities, or you could add `Loaded` to custom bundles of your entities. Or similarly, you could add `ToBeLoaded` at the time of spawning, and then the `added_loaded_component` system adds `Loaded` only to the entities `With<ToBeLoaded>`. Just avoid using a blanket solution because then you despawn everything with `teardown`. Hope it helps!

**Comment by Sycrosity** [comment]
Ah yes that was it - I had added a persist label to the camera entity but forgot about the window... I resolved this by in my startup system adding `Persist` to all entities, so that my system would work - I'm aware this is probably not the best way to do it but to avoid adding a component manually to every entity I want torn down, this seems to work. I'll mark this as answered - thanks for your help and your detailed explanation!
